liiatools/ssda903_pipeline/sufficiency_transform.py

```diff
@@ -655,6 +655,7 @@ def postcode_transform(df: pd.DataFrame) -> pd.DataFrame:
 
 def ofsted_transform(fs: FS, ONSArea: pd.DataFrame) -> (pd.DataFrame, pd.DataFrame):
     """Creates a dimension table with all Ofsted providers and a fact table with all Ofsted inspections"""
+    fs.makedir("Ofsted", recreate=True)
```

Comment:
This will create a new empty directory every time that it is run, erasing all previous data. 
If that is the behaviour that you want then this works.

Reply:
Not necessarily, but I can see why you'd understand that.  According to the source code, it just ignores any errors if set to true.  https://docs.pyfilesystem.org/en/latest/_modules/fs/base.html#FS.makedir

```
            dir_paths = tools.get_intermediate_dirs(self, path)
            for dir_path in dir_paths:
                try:
                    self.makedir(dir_path, permissions=permissions)
                except errors.DirectoryExists:
                    if not recreate:
                        raise
            try:
                self.makedir(path, permissions=permissions)
            except errors.DirectoryExists:
                if not recreate:
                    raise
            return self.opendir(path)
```

So, it seems that recreate is just a flag as to whether to raise an exception or not.  An alternative approach would be to allow it to error and then catch and handle the error. This may future-proof things.

Maybe something like this would be better?

```
try:
     fs.makedir("Ofsted")
except fs.errors.DirectoryExists:
     pass
except fs.errors.DirectoryExpected as err:
     log.error(f"File already exists and is not a directory: {err}")
```

Reply:
Thank you. That makes sense now. My reference was from the behaviour that I observed yesterday when [writing this](https://github.com/SocialFinanceDigitalLabs/sf-django-filesystem/blob/477ae438caa1a10e8b6b4cb111a92ef3a1830a76/filesystem/filesystem.py#L34). It is similar but not the same. 

Seeing makedir's code makes it clearer. 

This works then. Will approve.

Reply:
That makes sense and is structured more the way I'd like. So it's good to link in here.  Who knows if later they decide to add that behaviour in though as "recreate" implies exactly what you said. I'd rather handle this your way as opposed to relying on a hidden meaning though, so I agree with your comment.

Reply:
Two points here:
1. I've done some more testing and it seems like after the first time you run this extra line it's not longer needed. My guess is S3 will now recognise it as a directory, so in theory it just needs to be run once whenever it's referring to a new bucket. Not super helpful but just interesting (makes testing slightly harder as I now can't recreate the error) and we should of course still be checking for this error
2. We have used `recreate=Yes` in several parts of the pipeline already. Should we now go through and update all of those uses?

Reply:
1. Yeah, I guess it's trickier to test the behaviour on S3. Testing was slightly more straightforward for me because for the liia-frontend project, I was running it on local. So I could delete (fake) folders and run the command again to see if it detects and recreates the folders that are missing.
2. If you decide to, then maybe its worth creating a utils function (e.g `ensure_dir(path)`) that contains the try-except clause. So that the makedir line is replaced with an ensure_dir line instead of the full four lines each time. Redirecting the decision about this to @cyramic

Reply:
Great points about some of the aspects of working with S3.  Good to know for the future.

I didn't appreciate number of times this is in the code.  I only flagged it here as it seemed to cause an issue on my test, and thought it would be good to harden this code a bit so we can control and surface the issue a bit.  I agree with Tambe's suggestion in that if you can, spin it out to a separate function.   For now, however, I'm happy with you doing it just with this code, then add an item to the backlog about the other functions as tech debt. We can then prioritise it later to act upon.

Hope that makes sense!

Reply:
Have adjusted just this section now. I didn't include the DirectoryExpected error catching I don't believe that error will show up when trying to use fs.makedir